Technocrat Pagination fixes

NigelCunningham opened the pull request:
Hi there!

Since I submitted a PR for breaking tables across pages last year, I've continued to work on improvements for pagination in pagedJS. We've reached the point at which the project has gone live and we've worked the bugs out, so I'm now submitting the changes upstream for your consideration. I see you have a lot of issues open that relate to pagination. I'll link them to this one so the authors can become aware of this patch and give it a try. I'd also like to add unit tests if you're willing.

I don't imagine for a second that this patch handles every conceivable situation, but it does provide a big step forward and hopefully a good basis for future work.

The main changes are:

- Finds and removes multiple pieces of overflow at once. This deals with situations such as where you have a table with multiple columns, and additional rows below the one that needs breaking. It lets us remove the overflow from each column in the row that needs breaking, and also remove the rows that are completely off the page, all at once. Removing the overflow one at a time was problematic - it caused reflow on other columns that could then mean you ended up with lots of small pieces of overflow and confusing issues around the order in which they needed to be readded. Automatically handle table splits and inline elements as well as possible so that we're not too dependent on css classes for hints.
- Always render the highest level elements that have multiple children to completion before looking for overflow. This is part of the previous improvements.
- Ensure that if the overflow is itself too long for a page, it gets split too before we add anything else.
- Scroll to a newly added page as we add content - helps with debugging.
- Seek above all else to never get the loops the current code works to detect; we should always be making forward progress and shouldn't need the maxChars code. In this version it is pretty unused but not completely removed.
- Handle splitting of nested lists across pages too.
- Handle issues with fractional pixels; the rounding is not always a good idea.

julientaq commented:
oh that’s quite a big one.

Thanks for sharing the work. we’ll have @fchasen to dig this pr quickly.

fchasen commented:
Thanks @NigelCunningham! These changes look amazing - looking over them now but a lot to check out.

I'm hoping we can split these PR up a bit, so it's easier to merge and check the other current PRs against if you don't mind. Not sure where those boundaries should be yet hopefully should have some suggestions in the next few days.

NigelCunningham commented:
You're welcome!

Unfortunately a lot of the issues are interconnected and hard to split out by themselves. A good starting point might be adding the overflow class and the allowance for a breakpoint to include multiple pieces of overflow. I'd like to help but am really busy with lots of other things. That said, if you need explanation for anything, I'll do my best to help.

relaxatorium commented:
Hello all,

I'm having some issues rendering tables in our application that uses pagedjs as a big part of a PDF rendering pipeline (very happy with it overall!), so I tried this MR to see if I am about to be quite happy soon.

I ran into an error instead.

`Caught error:  Error [TypeError]: Cannot read properties of undefined (reading 'breakToken')
Page.layout (/Users/dlstern/Development/pdf_server/node_modules/pagedjs/dist/paged.polyfill.js:2622:37)`

In my PDF pipeline, or

`Unhandled Promise Rejection: TypeError: undefined is not an object (evaluating 'renderResult.breakToken')` when I simplified the example into an html page as attached here for repro purposes.

I'm not familiar enough at this time with the guts of this library to do more detailed debugging myself, but I have at least attached the html table that I used to generate this error when using the library version built out from this pull request.

[table-example.html.zip](https://github.com/pagedjs/pagedjs/files/13242278/table-example.html.zip)

Let me know if I can do anything else to help, I am excited to see pagination improvements for this great library, especially with tables where we've been having issues at times.

NigelCunningham commented:
Thanks @relaxatorium for the feedback. My apologies - I had some fixes from rebasing to 0.4.3 that I missed committing. Please pull the extra commit and try again.

relaxatorium commented:
That's completely alright, happy to take another look at this today and update you on the progress

relaxatorium commented:
Trying the latest version, I am no longer seeing the error mentioned above but I am also not seeing any pages after page 1

NigelCunningham commented:
Hmm, something strange is going on there. getBoundingClientRect(wrapper.children[3]) is returning that the height of the table is only 705.5 (and the width is 6868px!), but the Computed style in Chrome's debug tools is 700.891 x 2408.880. Very strange! (The calculations are saying the table isn't overflowing).

I need to head off to bed for now, but will hopefully find some time to look some more tomorrow evening.

janschreieradac commented:
Hi there,
I came here from Issue 114 and managed to test the PR. It solves all my issues :)
Thanks for working on this!

NigelCunningham commented:
That's great to hear, @janschreieradac!

Perhaps I should give an update, too: Since it looks like @relaxatorium's issue might be in Chrome itself, I've downloaded the browser source and succeeded yesterday in building from source. I had a quick little try just now at debugging, but will need to pick this up later (I need to get on with work). Progress, though!

raghuramvolteo commented:
@NigelCunningham , I have created the build(from branch `te/pagination`) and from the dist folder have taken the paged.polyfill.js. This is leading to showing of only two pages, all other pages are getting cropped(missing pages).  Please let me know if I should give more context?

But if I take the same file from the original build from main branch - it has rendering issue in the table, but pages are not getting cropped.

husstahn commented:
Tried build using node v16.20.0 v18.16.1 same result as @NigelCunningham

NigelCunningham commented:
I'm continuing to work on this and hope to have more commits soon that will address the issues found. Thanks for the feedback!

NigelCunningham commented:
Ok, I've got a workaround for the issue, which gets the HTML being rendered in @raghuramvolteo's sandbox and the @relaxatorium's sample table rendering fine. I'm not happy because I want to understand why the overflow goes to the right instead of below, but that's being handled. If possible, would you both retest yourselves and confirm/correct my assertion? Thanks!

NigelCunningham commented:
Great; thanks! I'started working through sample files provided with other issues too. One of them (https://github.com/pagedjs/pagedjs/issues/140) has a particularly complex table, so once I get that going we should be in a really good place.

relaxatorium commented:
A quick look at the test example project looks good, I'll get this in my full pipeline and test it there when I can shortly or early tomorrow, and report on that as well.

NigelCunningham commented:
Thanks! If you have any issues, please provide some sample HTML I can test with :)

relaxatorium commented:
This latest version that was pushed up was actually a major regression in my example project. The table stopped displaying and the page number went to X of 0 with the "X" rapidly incrementing, as shown in the attached screen recording

https://github.com/pagedjs/pagedjs/assets/411825/44755df2-d5c9-4ea6-8b27-104803fd2e94

NigelCunningham commented:
Hi Dave.

Yeah, I had it working perfectly earlier in the day, then made more changes for other issues. After pushing last night I retested this and found the same; I'm about to do a bit now before starting work; hopefully it will be a quick and easy fix.

NigelCunningham commented:
Ok; I've found the commit that broke the rendering for you, @relaxatorium. The other samples don't seem to be broken by reverting that single change so I've just pushed the revert. Please retry when you have the opportunity.

Thanks!

julientaq commented:
We’re looking into the branch to see how to merge this in our code base,  while making sure we understand all the code, that’s why @fchasen wanted to divide the request in multiple ones.

Since paged.js is in used in multiple places with custom hooks, we need to be careful we don’t break everything.

@NigelCunningham this is a great pull request, it’s just a tiny bit big :D, but we’re workin’ on it.

NigelCunningham commented:
Thanks @julientaq; I'm just focusing on getting it working with all use cases for now :)

NigelCunningham commented:
Hi again.

I now have the code working - perfectly I think - with all of sample HTML provided in related issues.

fchasen commented:
@NigelCunningham sorry again that this is taking a while to get merged - really appreciate these changes and I think by getting these in we will have figured out the process to allow the chunker part of Pagedjs to iterate faster. 

I've finally got the tests running in Github actions, so if you can rebase from main it should run on this PR now. 

When I run the specs locally with `npm run docker-specs` I'm seeing 172 failures. Many of those are probably minor shifts in rendering, but some like `specs/breaks/breaks.html` don't seem to be working with these changes.

NigelCunningham commented:
Thanks.

I'll seek to get onto this over the weekend.

Nij

On 15 December 2023 4:36:43 am AEDT, Fred Chasen ***@***.***> wrote:
>@NigelCunningham sorry again that this is taking a while to get merged - really appreciate these changes and I think by getting these in we will have figured out the process to allow the chunker part of Pagedjs to iterate faster. 
>
>I've finally got the tests running in Github actions, so if you can rebase from main it should run on this PR now. 
>
>When I run the specs locally with `npm run docker-specs` I'm seeing 172 failures. Many of those are probably minor shifts in rendering, but some like `specs/breaks/breaks.html` don't seem to be working with these changes.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/pagedjs/pagedjs/pull/171#issuecomment-1856288007
>You are receiving this because you were mentioned.
>
>Message ID: ***@***.***>

NigelCunningham commented:
I've got it rebased but can't push at the moment - my access to our repo is messed up. I'll push the update as soon as I'm able.

NigelCunningham commented:
Ok, the rebase is now pushed @fchasen.

NigelCunningham commented:
I'm now having a try at getting `npm run docker-specs` running so I can see if I can help with the issues there.

NigelCunningham commented:
I've started on fixing the test failures. One of the disparities is due to the fact that the new code is setting data-split-from where the old code didn't, resulting in a shift in the content (margin-top etc). I'll disable setting the split-from to make the tests pass but we might want to consider whether the which behaviour is correct/preferred.

almogyehuda commented:
There's any estimation to get this solution as a version?
Will be much appreciated thanks.

hendrikheil commented:
Just tried this branch on an existing complex project. It fixes things like missing rows when a table is chunked across two pages, but breaks multi column tables for us. Also `break-after: always;` doesnt seem to work on the chunked table anymore. So there wont be any whitespace after a table is chunked anymore, the following page is rendered immediately after the last row of the table. Not sure if this PR is supposed to also render the table head on each chunk, but it seems like currently it doesn't yet.

Happy to contribute or give more details!

NigelCunningham commented:
Thanks for the feedback. I'm currently working through the automated tests, addressing a bunch of regressions that were caused by the fact that I wasn't even aware of some features like footnotes and just got things working for our internal use case. I have a bunch of additional commits that I'm sitting on so I can not make too much of a mess of the branch. Please keep watching this space :)

NigelCunningham commented:
By the way, @hendrikheil , if you could provide me with a sample of your multi-column table HTML and what you're expecting, I'll include that in my testing too.

hendrikheil commented:
@NigelCunningham, I just created a little "fixture" of how I would expect the table with a columns layout to work. `expected.html` should probably just be used to see what I would have expected visually, as I just hacked a few styles and wrangled the markup. `output.html` is what I get from the input as of `10cbe5c` (last commit currently)

NigelCunningham commented:
I'm up to about 80 of the 90 odd tests passing, so not much more work to do. Apologies for the quietness!

hendrikheil commented:
Looks promising already! Really looking forward to the end result. Thanks @NigelCunningham!

julientaq commented:
yes, quite impressive.

NigelCunningham commented:
That's 95 of 95. Some of the tests needed to be updated, but I didn't cheat, I promise!

NigelCunningham commented:
@fchasen could I get you to take a look now?

I've fixed all of the issues in the existing tests (some tests needed updating; please check them - in some cases there were complaints about differences in the output but nothing was discernable so I updated the image to make the test pass; in at least one of the cases, I think the old behaviour was wrong so updated the test there too).

For the moment at lest, I consider this branch done. I do need to recheck against our own contracts to make sure I haven't broken them with all these changes, too.

NigelCunningham commented:
By the way, the issue with the overflow going to the right instead of down was related to the display:grid and use of columns on the page. Not a bug in Chrome at all, just not a no-brainer :)

almogyehuda commented:
I'm still able to reproduce the bug with the latest commit in this PR.
The table disappears instead of break between pages - only when I'm giving it overflow: auto I can see some portion of the table - but it's cutoff as probably understood.

fchasen commented:
Thanks again @NigelCunningham for doing the monumental task of updating all those specs to pass.
All tests are passing locally for me as well and diffs of the changed images all look good.

Should have this merged and released very soon!

fchasen commented on src/chunker/chunker.js:
```diff
@@ -332,48 +333,63 @@ class Chunker {
-					tokens.push(newToken);
-				}
+				// Debugging tool - ensures previously rendered content is visible.
+				window.scrollTo(0, document.body.scrollHeight);
```
If this is just for debugging, I think we should either remove this or put it behind a debug flag. 

When testing with the Moby Dick file, I'm seeing ~200ms faster rendering without this scrolling.

NigelCunningham replied:
Ok, I'll remove.

fchasen commented on src/chunker/chunker.js:
```diff
@@ -104,8 +103,10 @@ class Chunker {
 		this.hooks.layout = new Hook(this);
 		this.hooks.renderNode = new Hook(this);
 		this.hooks.layoutNode = new Hook(this);
+		this.hooks.getOverflow = new Hook(this);
```
Could you add these new hooks to the list in the Readme?

NigelCunningham replied:
Sure!

NigelCunningham replied:
Removed the hook now - it ended up being an unneeded addition.

fchasen commented on src/chunker/layout.js:
```diff
@@ -385,17 +440,25 @@ class Layout {
+				temp = container;
+			} else {
+				// Offset can be incorrect in tables spanning multiple pages.
+				// To reproduce, fill a specification composite "prepared by" field
```
Not clear what this means? Is this something we should add a new spec to test for?

NigelCunningham replied:
I don't remember that one - will have to look at it more.

NigelCunningham replied:
The comment was added by a co-worker. I tried removing the `|| child(container, 0)`, which they also added and it did result in 4 test failures, but in every case there was no noticable actual difference so I believe the issue they were addressing is already covered by other changes and tests.

Removing the modification they made.

fchasen commented on src/chunker/layout.js:
```diff
@@ -467,233 +530,572 @@ class Layout {
 		}
 
 		if (!node) {
+			// console.log(isElement(container), child(container, offset), temp);
```
Can probably remove this or add it for debugging if it's helpful?

NigelCunningham replied:
Will remove.

fchasen commented on src/chunker/layout.js:
```diff
@@ -467,233 +530,572 @@ class Layout {
+
+		let refId = parentElement.dataset['ref'];
+
+		if (['TR', 'math', 'P'].indexOf(parentElement.tagName) > -1 && parentElement.textContent.trim() == '') {
```
Can you add a comment on what this is doing?

Wondering if we need this list to have any other elements and if it is ok to have `math` lowercase.

NigelCunningham replied:
Yeah, the tagname for math was lowercase, which is unusual. Will add a comment and remind myself what's special about TR and P.

fchasen commented on src/chunker/layout.js:
```diff
@@ -467,233 +530,572 @@ class Layout {
+						} else {
+							// Has no children so should have no height, all other things
+							// being equal.
+							let childBounds = getBoundingClientRect(node);
+							intrinsicRight = getBoundingClientRect(node).right;
+							intrinsicBottom = getBoundingClientRect(node).top;
+							let intrinsicLeft = getBoundingClientRect(node).x;
```
Can we use `childBounds` instead of `getBoundingClientRect` again for the next few lines?

```suggestion
							let childBounds = getBoundingClientRect(node);
							intrinsicRight = childBounds.right;
							intrinsicBottom = childBounds.top;
							let intrinsicLeft = childBounds.x;
```

NigelCunningham replied:
Will check.

NigelCunningham replied:
When I actually read the code, there's nothing to "check". The answer should have been "yes, oops!". Done now!

fchasen commented on src/chunker/layout.js:
```diff
@@ -467,233 +530,572 @@ class Layout {
+						intrinsicRight += parseInt(styles["paddingRight"]) + parseInt(styles["marginRight"]);
+						intrinsicBottom += parseInt(styles["paddingBottom"]) + parseInt(styles["marginBottom"]);
+					} else {
+						let childBounds = getBoundingClientRect(node);
```
Can we also reuse `childBounds` here?

NigelCunningham replied:
Will review.

NigelCunningham replied:
Ditto here.

fchasen commented on src/utils/dom.js:
```diff
@@ -591,6 +788,11 @@ export function indexOfTextNode(node, parent) {
 	}
 	let nodeTextContent = node.textContent;
+	// Remove hyphenation if necessary.
+	if (nodeTextContent.substring(nodeTextContent.length - 1) == '-') {
```
Is this from the hyphens we are adding when text is split in `hyphenateAtBreak()`?

If so we should check for `"\u2011"` and might need to follow up to remove the whatever character is set in `hyphenGlyph` option.

NigelCunningham replied:
I'll look at this and the next one - need a reminder of the context.

NigelCunningham replied:
Fixed, using both.

fchasen commented on src/utils/dom.js:
```diff
@@ -591,6 +788,11 @@ export function indexOfTextNode(node, parent) {
+	// Remove hyphenation if necessary.
+	if (nodeTextContent.substring(nodeTextContent.length - 1) == '-') {
+		nodeTextContent = nodeTextContent.substring(0, nodeTextContent.length - 1);
+		console.log(nodeTextContent);
```
```suggestion
```

fchasen reviewed (approved):
A few questions and nits, but this is looking really great.

relaxatorium commented:
This is looking like it will solve my table rendering issues when it goes live, which I am excited about. One edge case I will mention is that we are using a Handler based on https://gitlab.coko.foundation/pagedjs/pagedjs/-/issues/84#note_77124 to render repeated table headers, and this appears to be breaking that handler.

I'm investigating potential rewrites of that old handler and would gladly figure out how to rewrite it if it means my main rendering problem is fixed, but I figured I'd flag it up in case it either raises issues that need to be resolved or if you just had any thoughts on how that would need to be rewritten.

NigelCunningham commented:
Just adding a note to say this is not forgotten - I'm hoping to get to it this afternoon or in the coming week.

leonardolb commented:
Hi! I'm having an issue with large html text inside a table row: page breakes and stops rendering all other ones.

does these PR solve this issue too?

NigelCunningham commented:
@relaxatorium , I've added support for repeating a table header, including a test. Just add a repeat-header attribute to the tbody (<tbody data-repeat-header>).

NigelCunningham commented:
Ok, everyone. Are there any other items I need to work on?

leonardolb commented:
> Ok, everyone. Are there any other items I need to work on?

Hi! Just my issue other there.

This is my issue: text inside merged cell is a HTML text.

When rendering, paged stop rendering all other pages.

With large HTML text inside table (breaking pages)
<img width="574" alt="image" src="https://github.com/pagedjs/pagedjs/assets/109328159/c42b1a58-59a1-47a3-a29d-534229210b56">


It works fine if I display text as plain text inside table.

Same text, but no html

<img width="554" alt="image" src="https://github.com/pagedjs/pagedjs/assets/109328159/9fd7ade6-2a13-4922-8875-1b36c3fab99c">

NigelCunningham commented:
Hi @leonardolb.

Would you please provide the complete HTML that pagedJS is trying to paginate? Without that, I'll be wasting time trying to reproduce the issue and perhaps solving the wrong problem.

Regards,

Nigel

leonardolb commented:
Sorry about that @NigelCunningham 


I've simplified my sample html a little bit, but you can still reproduce the issue.


[no_paged.txt](https://github.com/pagedjs/pagedjs/files/14475585/no_paged.txt)
[with_paged.txt](https://github.com/pagedjs/pagedjs/files/14475586/with_paged.txt)

I've attached both with and without paged. 

I'll be glad to help further!

julientaq commented:
> @relaxatorium , I've added support for repeating a table header, including a test. Just add a repeat-header attribute to the tbody (`<tbody data-repeat-header>`).
> 
> (Edit: I did it this way to make repeating the header optional. It would be nice if CSS had orphan / widow attributes for tables that we could leverage, but I read they're for block elements (which I understand to exclude tables ... right?))


We need to be careful with what the w3c says here, and how we should manage this.

This `data-repeat-header` is a great idea. We should be able to use it to create complex hooks to have more complex repeating table headers. It’s a great idea! I think we can leverage css by creating a `--table-header-repeat` property to offer this feature to authors.

fchasen commented:
Thanks so much all, this will be merged in to main shortly so please file new issues for any bugs so we can track them separately. 

These changes will definitely break some hooks, so we will publish it as beta initially while we workout any issues and communicating the breaking changes. I'll create a `v0.5-beta` tag to track them for now., but hopefully it should be short beta period.

@NigelCunningham I'm so sorry to ask this but could you revert the repeating table header changes?

If you don't mind making those a follow up PR for this, I think we could use a custom CSS property (like @julientaq mentioned above) to enable or disable it instead of the data attribute (or at-least we can discuss it there).

NigelCunningham commented:
Sure. I'll do that in the morning.

Thanks for the feedback!

NigelCunningham commented:
Ok @fchasen , I've force pushed removing the last 2 commits (I think that makes it tider than adding a revert commit, which have always caused me grief later when it's come to reverting the revert).

NigelCunningham commented:
Thanks @leonardolb . Looking now and will create a new PR when I have a fix, but will link it here for you to easily find it.

NigelCunningham commented:
@leonardolb Do I need the stylesheets?

NigelCunningham commented:
@fchasen let's go ahead and merge this; I'll try to pick up related tickets if there's still more I can help with.

NigelCunningham commented:
Oh sorry @fchasen , I missed your comment above about "making those a follow up PR for this, I think we could use a custom CSS property (like @julientaq mentioned above) to enable or disable it instead of the data attribute (or at-least we can discuss it there)." Will do. Not sure how to implement the custom CSS property but I'm sure I'll learn :)

leonardolb commented:
> @leonardolb Do I need the stylesheets?

Hi, i gu
[Archive.zip](https://github.com/pagedjs/pagedjs/files/14589962/Archive.zip)
ess stylesheet don't relate to this issue, but I've attached css files to help reproduce.

julientaq commented:
@NigelCunningham  the goal is to write down what would be the best way to define the repeating table headers in the css.

I would assume that we could wrtie something like:

```css
thead {
  --pagedjs-repeating-header:  repeat all;
}
```

and then define how this repeating header should be reused. So we could have `repeat once`, `repeat all`, or `no-repeat` 

Or maybe we should use the `position:sticky`. 

If it’s `sticky`  it should be reusable on all the pages, before any content. 


There is a lot of different ways to think about how we could/should handle that, but it should come with specs, and some examples and use cases. If those seems good enough, we could then share that with the w3c and push the boundaries of what we do forward. 


Let’s see when this is merged :D 

(again, quite impressed by the work you made).

fchasen commented:
This is merged in now and published to NPM as `pagedjs@0.5.0-beta.1` .

Thanks so much @NigelCunningham for all your work on getting this in and really appreciate everyone testing and finding issues. 

There is a `0.5.0-beta` tag now, so please tag any issues you find during the beta with that: https://github.com/pagedjs/pagedjs/labels/0.5.0-beta

NigelCunningham commented:
Awesome, thanks Fred!

EnochGao commented:
@NigelCunningham  I used the latest version and found that when I created the iframe tag, the home header disappeared [demo](https://github.com/EnochGao/html)

julientaq commented:
@EnochGao Since this is merged, you may want to create a new issue and label it 0.5.1beta.

I’m adding the label now.

This way we can move on to the testing part.

relaxatorium commented:
@NigelCunningham @fchasen @julientaq 

Hello all!

Besides other standard careful testing with our production pipeline, the remaining issue related to all of this for my company's use case is in regards to the repeated table headers, as mentioned upthread https://github.com/pagedjs/pagedjs/pull/171#issuecomment-1961815948

From what I can see, an initial solution was proposed and implemented, but was then removed out of this PR, which makes sense as this PR was already quite large and that is a distinct enough issue that it makes sense for it to have its own PR and discussion.

Has that been followed up on, and if so, where is that PR and/or issue such that I can monitor it? I have not been able to find it. If there is no issue for this yet and there should be, I would be happy to write it up as needed.

julientaq commented:
hey. 

Yep, we removed the automatic table header because we think it would need to be doable on a css basis. 

@NigelCunningham is still working on making the beta futureproof a bit, we found a bunch of bugs that need fixing. When that is done, i think we’ll make an extension to handle repeated table headers. (for which we’ll write more specs).

relaxatorium commented:
Understood, so for now, just sit tight and wait for a PR to appear after the beta?

julientaq commented:
yep, 
although, thinkin about it, i believe we can reuse some code we had before.
let me test those next week, and i’ll come back to you soon.

relaxatorium commented:
Thank you kindly. We've loved using this project and we will love it even more with improved table pagination and with a more elegant way to repeat headers!

cyhnkckali commented:
When do you plan to release it?

NigelCunningham commented:
As soon as it's ready to go, I'll push it to the github repository.

iSokrat commented:
@NigelCunningham any updates, when it'll available on npm?

NigelCunningham commented:
We're just working through other issues that have been found. Hopefully it's not far away from being merged and another beta tagged. Thanks for your interest!

NigelCunningham commented:
You are see the current state in the pull request 196 if you'd like to test it out.

iSokrat commented:
@NigelCunningham Thanks for the quick response! We've integrated pagedjs to generate large reports which may contain large tables and for us it's really crucial moment to have aligned table. Without this it's hard to read the tables, maybe, you have some proposal with quick fix how to align width of the columns for the table which is partially moved to the next page? And the second question - maybe, you have some knowledge - is it possible to move header of the table to the next page as well?

NigelCunningham commented:
The code as it stands should match columns widths.

Regarding moving the header to the next page, do you mean copying it, so it appears on each page? Assuming that's the case, there is code that does that but I believe I need to at least recheck it is still working after recent changes.

Hope that helps!

iSokrat commented:
@NigelCunningham On version pagedjs "0.4.3", for the first page we have "colgroup" which contains settings for widths of each column
<img width="253" alt="image" src="https://github.com/pagedjs/pagedjs/assets/13063442/3377bd0a-fc03-431e-9df1-84588787f410">
but for the next paged of the same table, it's not included there -> width is not the same as on the first page:
<img width="249" alt="image" src="https://github.com/pagedjs/pagedjs/assets/13063442/52a45da7-8fde-4947-8b65-34816da3e444">

NigelCunningham commented:
Sorry - I didn't see your question the first time. I'll take a look.

NigelCunningham commented:
Actually... could you provide me with sample HTML, and could I get you to open a separate issue for this? We're ending up with lots of little things and I'm loosing track of what still needs attention - that would help. Thanks!

iSokrat commented:
Maybe, later some piece of html will be provided. For other guys, we have found a workaround: we just manually copy it from table to table via hooks:

```ts
import { Handler } from "pagedjs";

export class SplitTableHandler extends Handler {
  colgroupsMap = new Map<string, HTMLTableColElement>();

  afterPageLayout(pageEl: HTMLElement) {
    const tables = pageEl.querySelectorAll("table");

    tables.forEach(table => {
      const splitFrom = table.getAttribute("data-split-from");

      if (
        splitFrom &&
        this.colgroupsMap.has(splitFrom) &&
        !table.querySelector("colgroup")
      ) {
        table.insertBefore(
          // copy colgroup from the first piece of the table
          // to make sure column widths are preserved
          (this.colgroupsMap.get(splitFrom) as HTMLTableColElement).cloneNode(
            true,
          ),
          table.firstChild,
        );

        return;
      }

      const colgroup = table.querySelector("colgroup");
      const dataRef = table.getAttribute("data-ref");

      if (colgroup && dataRef) {
        this.colgroupsMap.set(dataRef, colgroup);
      }
    });
  }
}

```
